logicalplan/merge_selects.go

```diff
@@ -0,0 +1,132 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
+package logicalplan
+
+import (
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+)
+
+// MergeSelectsOptimizer optimizes a binary expression where
+// one select is a superset of the other select.
+// For example, the expression:
+//
+//	metric{a="b", c="d"} / scalar(metric{a="b"}) becomes:
+//	Filter(c="d", metric{a="b"}) / scalar(metric{a="b"}).
+//
+// The engine can then cache the result of `metric{a="b"}`
+// and apply an additional filter for {c="d"}.
+type MergeSelectsOptimizer struct{}
+
+func (m MergeSelectsOptimizer) Optimize(expr parser.Expr) parser.Expr {
```

Comment:
` Optimize(expr parser.Expr) parser.Expr {`

We need to invest in good documentation to make it easy to optimize with just this interface (: